carbon-design-system/carbon
# feat(react-hooks): add useDebounce and useThrottle

This is a follow-up PR to feedback from: https://github.com/carbon-design-system/carbon/pull/3907, namely that we need a way to debounce/throttle updates much like we would attach debounce/throttle to window events that we create.

One challenge with hooks is that using one means that you opt-in to its behavior, alongside any hooks that it itself uses. The majority of the time, this won't cause any concern. However, certain hooks may result in frequent updates and so you want to throttle or debounce the value that this hook provides. A naive example of this would be:

```jsx
function useNoisyHook() {
  const [value, updateValue] = useState(0);

  useEffect(() => 
    const intervalId = window.setInterval(() => {
      updateValue(value => value + 1);
    }, 10);
    return () => window.clearInterval(intervalId);
  }, []);

  return value;
}

function DemoComponent() {
  const value = useNoisyHook();
  return <span>{value}</span>;
}
```

If a component decides to include `useNoisyHook`, then it in turn will be called every 10ms when the interval is called. This in and of itself is not a problem, but if this in turn causes issues with the resulting layout or adds work to reconciliation that is undesirable, then you may want a way to "throttle" or "debounce" these updates so that the UI isn't updating as frequently.

For example:

```jsx
function DemoComponent() {
  const value = useNoisyHook();
  const debouncedValue = useDebounce(value, 1000);
  return <span>{debouncedValue}</span>;
}
```

In this example, `useDebounce` will defer changing `debouncedValue` until 1000ms after `value` has stopped updating. When used in the return statement, that means that the result will not change until this time period has been settled.

In practice, this hook will be valuable for handling noisy updates from window events and provides a base for building `useThrottle` as an alternative to dealing with noisy hooks.

#### Changelog

**New**

- `useDebounce` and `useThrottle` in `react-hooks`

**Changed**

**Removed**

#### Testing / Reviewing

- Pull down PR and run storybook by doing `yarn develop`
- Verify storybook story for `useDebounce` operates as expected (should mirror behavior one would receive with lodash)
